### PR TITLE
refactor(query): csv reader support prefetch

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- Fixes https://github.com/datafuselabs/databend/issues/14973

so for a large file read of next batch and processing of the prev can overlap.
make the whole time shorter especially for relative slow storage.

add PrefetchAsyncSourcer,  keep return `Async` (instead of `NEED_CONSUME`) until prefetch goal is reached or finished.
rely on some implementation details (correct me if  I am wrong @zhang2014 ): trigger of downstream processor dependent on `push_data`, not `NEED_CONSUME` event,  `NEED_CONSUME`  is  only used to change processor state to `IDLE`

### update

the reason for https://github.com/datafuselabs/databend/issues/14973 is simple: the num of processors is not correctly set.

fixed.

but the prefetching still make sense on slow backend. 

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14983)

### CODE REVIEW

**Comment by BohuTANG** [comment]
I have test this PR, it seems the csv copy speed from 0.06M/s to 0.09M/s, but still slow than the tsv(0.16M/s):
image

Is this make sense?

**Comment by youngsofun** [comment]
> I have test this PR, the csv copy speed from 0.06M/s to 0.09M/s, but still slow than the tsv(0.16M/s):
> 
> image
> 
> 
> 
> It seems `rows/s` for the CSV is not correct, now the COPY of CSV cost is same as TSV now, about 10minutes.
> 
> 

how many rows in this file？

**Comment by BohuTANG** [comment]
Update:
CSV COPY is also slow than TSV:
image


TSV will cost 9 minute:
image

**Comment by BohuTANG** [comment]
It seems this PR is not help for CSV:
image

**Comment by BohuTANG** [comment]
> * copy tsv.gz 3X faster than csv.gz #14973

You can try follow this issue: https://github.com/datafuselabs/databend/issues/14973

**Comment by BohuTANG** [comment]
> how many rows in this file？

You can try follow this issue: https://github.com/datafuselabs/databend/issues/14973

**Comment by youngsofun** [comment]
there 2 reasons for this pr not working:

1. prefetch is not working
2.  prefetch is or not working, but not the real reason for slow.

---

logging shows that prefetch is not working.  source still only run async only once after the downstream pull. cc @zhang2014,     any workaround?

**Comment by youngsofun** [comment]
@zhang2014 description updated. plz review again.

File: src/query/pipeline/sources/src/prefetch_async_source.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+        if self.is_inner_finish || self.inner.is_full(&self.generated_data) {
+            Ok(Event::NeedConsume)
+        } else {
+            Ok(Event::Async)
+        }
+    }
+
+    fn un_reacted(&self, _cause: EventCause, _id: usize) -> Result<()> {
+        if let EventCause::Output(_output) = _cause {
```
Comment (zhang2014) [inline src/query/pipeline/sources/src/prefetch_async_source.rs:108]:
```suggestion
        if let EventCause::Output(output) = cause {
```

File: src/query/pipeline/sources/src/prefetch_async_source.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        if self.is_inner_finish || self.inner.is_full(&self.generated_data) {
+            Ok(Event::NeedConsume)
+        } else {
+            Ok(Event::Async)
+        }
+    }
+
+    fn un_reacted(&self, cause: EventCause, _id: usize) -> Result<()> {
+        if let EventCause::Output(output) = cause {
```
Comment (youngsofun) [inline src/query/pipeline/sources/src/prefetch_async_source.rs:108]:
```suggestion
        if let EventCause::Output(_) = cause {
```

**Comment by youngsofun** [comment]
@BohuTANG  

now cost 11m36, much faster than before (24 - 32min), but still slower than expected 9min12 (my last run with `enable_new_copy_for_text_formats=0`)

this image is right before ending, the progress is correct now (99997497 rows,  15.47G)

 
image

**Comment by BohuTANG** [comment]
> now cost 11m36,
I think the cost time is ok.

Two questions:
1. Why we should do this setting: `enable_new_copy_for_text_formats=0`, does it mean the new copy impl is still need to improve?
2. This PR is helpful but is one part of the improvement(Means we can merge this PR first)?

**Comment by youngsofun** [comment]
> Why we should do this setting: enable_new_copy_for_text_formats=0, does it mean the new copy impl is still need to improve?

the new impl is much better to maintain, logic is infact the same, but nearly half of the codes are rewritten, I`m not sure there are no problem.  


> This PR is helpful but is one part of the improvement(Means we can merge this PR first)?

yes, we can merge first 


---- 

it is strange that the `Duraition` in the history is diff from the time. it is longer :

image

if it is not acceptable, we can set enable_new_copy_for_text_formats=0 by default until it is improved.

**Comment by BohuTANG** [comment]
> > Why we should do this setting: enable_new_copy_for_text_formats=0, does it mean the new copy impl is still need to improve?
> 
> the new impl is much better to maintain, logic is infact the same, but nearly half of the codes are rewritten, I`m not sure there are no problem.
> 
> > This PR is helpful but is one part of the improvement(Means we can merge this PR first)?
> 
> yes, we can merge first
> 
> it is strange that the `Duraition` in the history is diff from the time. it is longer :

This is the real time which cost.

> image
> 
> if it is not acceptable, we can set enable_new_copy_for_text_formats=0 by default until it is improved.

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**